Filter expression on a repeat crashes attempting to check applicability of Entities filter strategy

#### ODK Collect version
the master version 9d5e3aeb7b4f98d321532221572488afc38f0812
#### Android version
10, 14
#### Device used
Redmi 9T, Pixel 7a
#### Problem description
Error evaluating a field in a form.
![countFamily](https://github.com/user-attachments/assets/51cb6a6f-c8fd-4684-8adc-576c9dbd2aae)

(The issue doesn’t occur in the store version 2024.2.4)
#### Steps to reproduce the problem
1. Open the form.
[familymemberscount(1).xlsx](https://github.com/user-attachments/files/16781541/familymemberscount.1.xlsx)
#### Expected behavior
In the previous versions of Collect the worked without the error.

```
  Caused by: java.lang.NullPointerException: getInstanceId(...) must not be null
                    	at org.odk.collect.entities.javarosa.filter.LocalEntitiesFilterStrategy.filter(LocalEntitiesFilterStrategy.kt:35)
                    	at org.javarosa.core.model.condition.EvaluationContext.filterWithPredicate(EvaluationContext.java:382)
                    	at org.javarosa.core.model.condition.EvaluationContext.lambda$filterWithPredicate$0$org-javarosa-core-model-condition-EvaluationContext(EvaluationContext.java:383)
                    	at org.javarosa.core.model.condition.EvaluationContext$$ExternalSyntheticLambda0.get(Unknown Source:14)
                    	at org.javarosa.core.model.IdempotentExpressionCacheFilterStrategy.filter(IdempotentExpressionCacheFilterStrategy.java:32)
                    	at org.javarosa.core.model.condition.EvaluationContext.filterWithPredicate(EvaluationContext.java:382)
                    	at org.javarosa.core.model.condition.EvaluationContext.filterWithPredicate(EvaluationContext.java:377)
                    	at org.javarosa.core.model.condition.EvaluationContext.expandReferenceAccumulator(EvaluationContext.java:343)
                    	at org.javarosa.core.model.condition.EvaluationContext.expandReference(EvaluationContext.java:250)
                    	at org.javarosa.core.model.condition.EvaluationContext.expandReference(EvaluationContext.java:219)
                    	at org.javarosa.xpath.expr.XPathPathExprEval.eval(XPathPathExprEval.java:34)
                    	at org.javarosa.xpath.expr.XPathPathExpr.eval(XPathPathExpr.java:213)
                    	at org.javarosa.xpath.expr.XPathPathExpr.eval(XPathPathExpr.java:60)
                    	at org.javarosa.xpath.expr.XPathFuncExpr.eval(XPathFuncExpr.java:187)
                    	at org.javarosa.xpath.XPathConditional.evalRaw(XPathConditional.java:72)
                    	at org.javarosa.core.model.condition.Recalculate.eval(Recalculate.java:50)
                    	at org.javarosa.core.model.condition.Triggerable.apply(Triggerable.java:119)
                    	at org.javarosa.core.model.QuickTriggerable.apply(QuickTriggerable.java:41)
                    	at org.javarosa.core.model.TriggerableDag.evaluateTriggerable(TriggerableDag.java:566)
                    	at org.javarosa.core.model.TriggerableDag.doEvaluateTriggerables(TriggerableDag.java:542) 
                    	at org.javarosa.core.model.TriggerableDag.initializeTriggerables(TriggerableDag.java:465) 
                    	at org.javarosa.core.model.TriggerableDag.initializeTriggerables(TriggerableDag.java:448) 
                    	at org.javarosa.core.model.FormDef.initializeTriggerables(FormDef.java:710) 
                    	at org.javarosa.core.model.FormDef.initialize(FormDef.java:1114) 
                    	at org.odk.collect.android.tasks.FormLoaderTask.initializeForm(FormLoaderTask.java:407) 
                    	at org.odk.collect.android.tasks.FormLoaderTask.doInBackground(FormLoaderTask.java:236) 
                    	at org.odk.collect.android.tasks.FormLoaderTask.doInBackground(FormLoaderTask.java:80) 
                    	at org.odk.collect.async.SchedulerAsyncTaskMimic.execute$lambda$0(SchedulerAsyncTaskMimic.kt:37) 
                    	at org.odk.collect.async.SchedulerAsyncTaskMimic.$r8$lambda$91Ccusi_zfxwouZjauTWIiyE2dg(Unknown Source:0) 
                    	at org.odk.collect.async.SchedulerAsyncTaskMimic$$ExternalSyntheticLambda0.get(Unknown Source:4) 
                    	at org.odk.collect.async.CoroutineScheduler$immediate$1$result$1.invokeSuspend(CoroutineScheduler.kt:18) 
                    	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33) 
                    	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:104) 
                    	at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:111) 
                    	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:99) 
                    	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584) 
                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:811) 
                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:715) 
                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:702) 
```

--- a/entities/src/main/java/org/odk/collect/entities/javarosa/filter/LocalEntitiesFilterStrategy.kt
+++ b/entities/src/main/java/org/odk/collect/entities/javarosa/filter/LocalEntitiesFilterStrategy.kt
@@ -32,7 +32,7 @@ class LocalEntitiesFilterStrategy(entitiesRepository: EntitiesRepository) :
         evaluationContext: EvaluationContext,
         next: Supplier<MutableList<TreeReference>>
     ): List<TreeReference> {
-        if (!dataAdapter.supportsInstance(sourceInstance.instanceId)) {
+        if (sourceInstance.instanceId == null || !dataAdapter.supportsInstance(sourceInstance.instanceId)) {
             return next.get()
         }
 

--- a/entities/src/test/java/org/odk/collect/entities/javarosa/filter/LocalEntitiesFilterStrategyTest.kt
+++ b/entities/src/test/java/org/odk/collect/entities/javarosa/filter/LocalEntitiesFilterStrategyTest.kt
@@ -6,6 +6,7 @@ import org.hamcrest.Matchers.equalTo
 import org.javarosa.core.model.FormDef
 import org.javarosa.core.model.condition.EvaluationContext
 import org.javarosa.core.model.condition.FilterStrategy
+import org.javarosa.core.model.data.IntegerData
 import org.javarosa.core.model.data.StringData
 import org.javarosa.core.model.instance.DataInstance
 import org.javarosa.core.model.instance.TreeElement
@@ -14,6 +15,7 @@ import org.javarosa.form.api.FormEntryController
 import org.javarosa.form.api.FormEntryModel
 import org.javarosa.test.BindBuilderXFormsElement.bind
 import org.javarosa.test.Scenario
+import org.javarosa.test.XFormsElement
 import org.javarosa.test.XFormsElement.body
 import org.javarosa.test.XFormsElement.head
 import org.javarosa.test.XFormsElement.html
@@ -268,6 +270,44 @@ class LocalEntitiesFilterStrategyTest {
         )
 
         assertThat(scenario.answerOf<StringData>("/data/calculate").value, equalTo("Thing"))
+    }
+
+    @Test
+    fun `works correctly with filtering on a repeat`() {
+        val scenario = Scenario.init(
+            "Count people underage",
+            html(
+                head(
+                    title("Count people underage"),
+                    model(
+                        mainInstance(
+                            t(
+                                "data id=\"count_people_underage\"",
+                                t("people",
+                                    t("name"),
+                                    t("age")
+                                ),
+                                t("total_underage")
+                            )
+                        ),
+                        bind("/data/people/name").type("string"),
+                        bind("/data/people/age").type("int"),
+                        bind("/data/question").type("string"),
+                        bind("/data/total_underage").type("string").calculate("count( /data/people [age&lt;18])")
+                    )
+                ),
+                body(
+                    XFormsElement.repeat("/data/people",
+                        input("/data/people/name"),
+                        input("/data/people/age")
+                    ),
+                    input("/data/total_underage")
+                )
+            ),
+            controllerSupplier
+        )
+
+        assertThat(scenario.answerOf<IntegerData>("/data/total_underage").value, equalTo(0))
     }
 
     @Test

